text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,136 @@
+# Dynamic size region
+
+- RFC PR: https://github.com/tikv/rfcs/pull/0082
+- Tracking Issue: https://github.com/tikv/tikv/issues/11515
+
+## Summary
+
+Make the size of a region dynamic, and only set a upper limit.
+
+This is the first step that we try to support PiB scale cluster.
+
+## Motivation
+
+We have observed a lot of regressions when the count of regions increases. The major drawback
+comes from three aspects:
+
+1. Transactions need to access more regions, which lead to too many RPCs and have high latency;
+2. A node needs to drive more regions, which can have high resource usages and hurt performance;
+3. Tools and service like CDC/BR/GC need to loop over all regions, which is slow.
+
+We have hibernate raft (#23) to get around the most significant problem 2. But it's not always
+working as expected:
+
+- Workload with random access can still activate a lot of regions;
+- Hibernated region needs extra time to renew its lease, which lead to high tail latency;
+- It complicates the design of CDC/BR, which requires access all replicas instead of just leader now;
+- GC will still wake up all regions and cause periodical high usage.
+
+There are two source of region creations: size/keys split and table split. Table split is disabled by
+default in TiKV. Even if it splits a lot of regions, small one will be merged later.
+
+Size/keys split contributes the most count of regions. The current size of a region is 96MiB by default,
+and the key count is 960,000. They are set based on an assumption that region is the concurrent unit of
+requests. Too large can lead to smaller concurrency and high latency. Since keys is related to the size,
+we only discuss region size here.
+
+## Detailed design
+
+To reduce the count of regions, we can set the size of region to a larger value. More precisely, instead
+of 96MiB, we set the max region size to 10GiB (before compression). To get around the problem of
+concurrency, we introduce new concept bucket that is smaller ranges within a region. To make hotspot be
+recoginized and scheduled quickly, we choose the make the size of a region dynamic.
+
+![dynamic size](../media/dynamic-size.png)
+
+### Dynamic size
+
+The hotter a region is, the smaller its size becomes. To make it simplified, we choose 512MiB for hot regions
+and 10GiB for cold regions. So there are two types of split, hotspot split and general size split. Hotspot
+split is triggered by PD. General size split is triggered by TiKV itself.
+
+10GiB is just an example, it's allowed to change to a bigger or smaller value.
+
+### Bucket
+
+A region is split into several buckets logically. We will collect query stats by buckets and report the bucket
+to PD. For hotspot regions, buckets are split by scan. For cold regions, buckets are split by approximate
+range size. Every bucket should have the size about 128MiB. Their ranges and stats are all reported to PD. PD
```

Comment:
If all the bucket stats are reported to PD, it would consume a lot of memory.
We only need to report bucket stats when read/write workload exceed a threshold.

So for most of the regions, we only need to report the range of buckets.
The total memory on PD would be greatly reduced.

Reply:
The bucket size is about 128MiB, which means it can consume as much as the memory current small regions use. As far as I know, it's not an issue yet.

Reply:
@coocood Good point. Do you know what's the roughly memory consumption per region in PD today?
For 128MB bucket and 1PB total database size, it's about 800 million buckets.

Reply:
@BusyJay For bucket stats report, will both leader and follower need to report their buckets? And I guess it's very likely for the same region, the bucket range may be different. If only leader reports the buckets stats, then PD will have to add the follower buckets stats when calculating the machine level stats, or report machine level stats separately.

Reply:
Currently, the small region will report statistics on both the leader and follower. the TopN(1000) region of this node will be reported not all.  may buckets can keep the same behavior.

Reply:
> ...the bucket range may be different.

It doesn't matter. Only leader's buckets are used and follower's stats will be merged into leaders using similar algorithm described in flow report section.

> the TopN(1000) hot region's stats of this node will be reported not all.

I'm worried about accuracy and message layout. I prefer to report all metrics and do filter on the PD side. Let's see how it performs first.